src/index.js

```diff
@@ -809,7 +819,8 @@ class Firestore extends commonGrpc.Service {
    * @param {Stream} resultStream - The Node stream to monitor.
    * @param {Object=} request - If specified, the request that should be written
    * to the stream after it opened.
-   * @returns {Promise.<Stream>} The given Stream once it is considered healthy.
+   * @returns {Promise.<Firestore~StreamCallback>} A Promise that resolves with
+   * a callback that returns the resulting stream.
```

Comment:
This is pretty confusing.  We have an asynchronous method that returns a callback/factory function that you have to call to actually get the result of the asynchronous method?

I kind of understand how we ended up here, but are there other ways to structure this code perhaps?  I'm somewhat disinclined to figure out exactly what's going on myself, but offhand I'm wondering if this method should be asynchronous in the first place?  Why do we want to handle "error before the first byte is read" as a failed Promise but any failure after that point as a stream error?  If this method could just synchronously return a (wrapped?) stream, that would avoid the need for the StreamCallback thing, I think.

Feel free to push back, but I think it's worth thinking a bit about other potential ways to structure this since this is getting a bit wonky.

Reply:
As per Josh's suggestion, I can rely on Node's implementation of microtask and macrotasks and schedule a macrotask here to make sure that the Promise chain gets resolved first.

Oh, and I really really really hated the original approach, but it was the only one I could come up with that at least somehow kept the original API contract: readStream() and readWriteStream() return a Stream that is known to be writeable.